PHP 7 on Travis-CI (re-redux)

weierophinney opened the pull request:
This is a version of #7270 (which was in turn a rebased version of #7263), rebased against master, to support PHP 7.

I have tested both against a nightly version from 2 weeks ago, as well as the Coercive STH patch. The latterwhich provided me with a number of issues in the ZF2 codebase with regard to passing invalid arguments to internal functions, and the bulk of the commits in this branch address those, keeping compatibility with PHP 5 versions as well as PHP 7 nightly.

There is one issue I've found in `Zend\Stdlib\ArrayUtils::filter()`: apparently, PHP 7 nightly has a regression in its `array_filter` implementation; the third argument, which indicates whether to pass values only, keys only, or both to the composed callback, is ignored, causing one test case to fail.

Otherwise, this version should be ready to merge anytime.

samsonasik committed:
add php 7 on .travis.yml

samsonasik committed:
update script for xdebug.ini check

weierophinney committed:
Disable apcu, redis, and memcached features on PHP 7 builds

weierophinney committed:
Broaden test assertion

We're ensuring at least one item is in $_ENV, but we are not resetting
its state. As such, the comparison should be that it has at least 1
item, not exactly 1.

weierophinney committed:
Rename duplicated argument to prevent error

weierophinney committed:
Disable a test that cannot run on v7

weierophinney committed:
Fix expected exception type

weierophinney committed:
Ensure session save handler stub has appropriate return values

weierophinney committed:
Better tests for HHVM/PHP7 for conditional travis installs

weierophinney committed:
End conditional statements in travis config

weierophinney committed:
Disable cache-specific travis integrations

weierophinney committed:
Fixed argument redefinition problems

- Prior to PHP 7, you could specify the same name for multiple arguments; PHP 7
  makes this a fatal error. The "feature" was used in several tests where 2 or
  more arguments from a data provider were not being used. Updated these to use
  discrete names for each argument.

weierophinney committed:
Fix E_RECOVERABLE_ERROR detection in Zend\Di tests

Prior to PHP 7, an E_ERROR was returned when trying to instantiate a
class but omitting required arguments; with PHP 7, this is now an
E_RECOVERABLE_ERROR. As such, trapping such errors intentionally
requires an error handler.

weierophinney committed:
Skip some Logger tests when under PHP 7

Both calling an undefined instance method as well as undefined static method now
raise an E_ERROR (instead of an E_RECOVERABLE_ERROR), which means shutdown
functions cannot be called... nor can the tests be executed.

weierophinney committed:
Ensure height and width are integers

when used as parameters to various `image*()` functions

weierophinney committed:
Cast data to string before caching

Ensure the data being written is a string before attempting to write it.

Also, `usleep()` requires an integer; change made to allow Filesystem storage
tests to pass.

weierophinney committed:
ob_implicit_flush expects an integer; dba_replace, a string

Even though the ob_implicit_flush docs use the boolean constants, it actually
expects an integer, either 0 or 1. Because PHP.

dba_replace expects string data.

weierophinney committed:
Ensure fgets() returns a string before calling explode

weierophinney committed:
Ensure we have a string before attempting to trim()

weierophinney committed:
Use string value when performing strtr

Oddly, there was a cast version of the value already; it just wasn't being used
in strtr().

weierophinney committed:
Ensure we have a string before calling rtrim()

weierophinney committed:
Ensure we have a string to pass to base64_decode

weierophinney committed:
Ensure strlen gets a string

weierophinney committed:
Ensure strtolower and addcslashes receive string arguments

weierophinney committed:
Skip tests that cannot run in PHP 7

Error that the test was trapping has been converted to an E_FATAL starting in
PHP 7.

weierophinney committed:
importNode() expects integer for argument 2 starting in PHP 7

weierophinney committed:
DateTime::diff's second argument is actually an integer

weierophinney committed:
Ensure substr receives an integer 3rd parameter

weierophinney committed:
Ensure preg_match receives a string value

weierophinney committed:
DateTime::diff's second argument is actually an integer

weierophinney committed:
Ensure substr receives an integer 3rd argument

weierophinney committed:
Ensure strtolower receives a string

weierophinney committed:
Ensure strlen and strpos receive strings

weierophinney committed:
Ensure substr receives an integer offset

weierophinney committed:
No need to do substr() if start is 0 and no length provided

weierophinney committed:
Ensure explode receives a string 2nd argument

weierophinney committed:
Ensure explode receives a string 2nd argument

weierophinney committed:
Skip tests in PHP 7 that we cannot run

weierophinney committed:
Ensure class_exists receives a string

weierophinney committed:
Override MongoDBOptions::__set to validate arguments

- Instead of relying on PHP errors, as invalid types raise E_FATAL

weierophinney committed:
ini_set expects a string as the 2nd argument

weierophinney committed:
Ensure we have a version string for version_compare

weierophinney committed:
Ensure ini_set receives a string value

weierophinney committed:
Ensure substr receives an integer 2nd parameter

weierophinney committed:
Ensure sscanf receives a string

weierophinney committed:
Ensure substr receives an integer offset, and preg_match receives a string value

weierophinney committed:
Updated magic.mime file for PHP 7 testing

weierophinney committed:
Ensure strlen receives a string

weierophinney committed:
Ensure DOMDocument::createElementNS receives a string 3rd argument

weierophinney committed:
Ensure imagefilledrectangle receives integers for offset positions

weierophinney committed:
Ensure addcslashes receives a string value

weierophinney committed:
Ensure idn_to_ascii returns a string

weierophinney committed:
Ensure we have reasonable dependencies

- Used semantic operator for all deps.
- Added composer.lock to repo, to speed up builds.

weierophinney committed:
Rename Int filter to ToInt

"int" is a reserved keyword starting in PHP 7. This patch changes the
class name to `ToInt`. A later patch will provide the original class,
but have it extend the new one.

weierophinney committed:
Re-add Int filter for BC

Re-adds the `Int` filter for backwards compatibility. The new version
extends `ToInt`, but raises an `E_USER_DEPRECATED` in the constructor,
instructing users to use the new filter.

weierophinney committed:
Renamed Int validator to IsInt

PHP 7 reserves "int" as a keyword, necessitating a change in the class
name; a later commit will re-add the original class, with a deprecation
notice.

weierophinney committed:
Re-added Int validator for BC

This patch re-adds the `Int` validator to retain backwards compatility. It
extends the `IsInt` validator, raising an `E_USER_DEPRECATED` error
during instantiation.

weierophinney committed:
Renamed Float validator to IsFloat

PHP 7 marks "float" as a reserved keyword; this patch renames the
`Float` validator to `IsFloat` for forward compatibility. A later patch
will re-add the original class with a deprecation notice in the
constructor.

weierophinney committed:
Re-add Float validator for BC

This patch re-adds the `Float` validator for backwards compatibility. It
extends the `IsFloat` validator and raises an `E_USER_DEPRECATED` notice
during instantiation to warn users to migrate.

weierophinney committed:
Renamed Float DDL column type to Floating

PHP 7 reserves "float" as a keyword; this patch renames the class to
`Floating`. A later patch will reintroduce the original class for
backwards compatibility, and raise a deprecation notice.

weierophinney committed:
Re-add Float column for backwards compatibility

Re-added the Float DDL column class to retain BC; it now raises an
`E_USER_DEPRECATED` notice on instantiation to warn users to migrate.

weierophinney committed:
Renamed String XML-RPC value to Text

"string" is now a reserved keyword in PHP 7, so the class must be
renamed. Updated `AbstractValue` to use the new class name in all
instances, ensuring that both the Server and Client will continue to
work seamlessly.

weierophinney committed:
Updated various XmlRpc tests to s/String/Text/

weierophinney committed:
Reference ToInt filter in Console tests

weierophinney committed:
Fix failing Form tests

- Range element uses a float validator internally; updated it to use `IsFloat`
- A number of tests and test assets used the `Float` validator; since they do
  not use the plugin manager, these were updated to use `IsFloat` instead.

weierophinney commented:
Only failure on PHP 7 at this point is with `Zend\Stdlib\ArrayUtils::filter()`, which is due to a regression in PHP 7's `array_filter()` implementation, which I have reported here: https://bugs.php.net/bug.php?id=69299

weierophinney committed:
CS fixes reported by php-cs-fixer

- Unused use
- whitespacey lines

weierophinney commented:
This change will require some documentation:

- `Zend\Filter\Int` is deprecated (`int` is a reserved keyword in PHP 7); `Zend\Filter\ToInt` is preferred.
- `Zend\I18n\Validator\Int` is deprecated; `Zend\I18n\Validator\IsInt` is preferred.
- `Zend\I18n\Validator\Float` is deprecated (`float` is a reserved keyword in PHP 7); `Zend\I18n\Valdidator\IsFloat` is preferred.
- `Zend\Db\Sql\Ddl\Column\Float` is deprecated; `Zend\Db\Sql\Ddl\Column\Floating` is preferred.
- `Zend\XmlRpc\Value\String` has been renamed to `Zend\XmlRpc\Value\Text` (`string` is a reserved keyword in PHP 7). This should only affect developers who are directly instantiating XML-RPC value objects, which is so uncommon as to not warrant a BC shim.

For filters and validators, if you are pulling them from the plugin managers, you will continue to be able to do so under the old names, and the new class will be resolved; if you are instantiating them directly, you will receive a deprecation notice indicating the new class to use.

DASPRiD commented on library/Zend/Cache/Storage/Adapter/Dba.php:
```diff
@@ -377,8 +378,10 @@ protected function internalSetItem(& $normalizedKey, & $value)
+
         $this->_open();
-        if (!dba_replace($internalKey, $value, $this->handle)) {
+        if (!dba_replace($internalKey, $cacheableValue, $this->handle)) {
```
This variable is only used once, why not cast it in-place?

weierophinney replied:
So I could provide a comment as to why the casting was happening.

weierophinney commented:
all is well! As @DASPRiD has reviewed, I'll merge, and push an RC3 immediately.

weierophinney commented:
So, @DASPRiD pointed out two more PHP RFCs that would affect this:

- https://wiki.php.net/rfc/reserve_more_types_in_php_7
- https://wiki.php.net/rfc/reserve_even_more_types_in_php_7

I found 5 more classes that will need shims for forward and backwards compat:

- `Zend\Form\Annotation\Object`
- `Zend\Filter\Null`
- `Zend\Paginator\Adapter\Null`
- `Zend\Mail\Transport\Null`

I'll work on these now. Until those are ready, I'm marking this as WIP.

weierophinney committed:
Rename Object form annotation to Instance

`object` will be a reserved word in PHP 7, and, as such, we cannot use it as an
annotation class name. `Instance` was chosen, and `Object` has been altered to
extend `Instance` and emit an `E_USER_DEPRECATED` notice on instantiation.

weierophinney committed:
Renamed Null log writer to Noop

`null` will be a reserved keyword in PHP 7; this patch renames the `Null` log
writer to `Noop` (as it performs a no-op). A later patch will re-define the
`Null` writer to extend `Noop` and emit a deprecation notice.

weierophinney committed:
Re-add Null log writer for BC

This patch reinstates the Null log writer for BC purposes. It extends
Noop, and emits an `E_USER_DEPRECATED` notice on instantiation.

weierophinney committed:
Renamed Null filter to ToNull

`null` will be a reserved keyword in PHP 7. This patch renames the `Null` filter
to `ToNull`, and ensures the filter plugin manager will use the new class. A
future patch will reinstate the `Null` filter with an `E_USER_DEPRECATED`
notice.

weierophinney committed:
Re-added Null filter for BC

Re-added the Null filter for BC purposes; it now emits an `E_USER_DEPRECATED`
notice on instantiation, warning the user to migrate.

weierophinney committed:
Renamed Null pagination adatper to NullFill

`null` will be a reserved keyword in PHP 7. This patch renames the
`Null` pagination adapter to `NullFill` (which describes better what it
does). A future patch will reinstate the original, but emit an
`E_USER_DEPRECATED` notice.

weierophinney committed:
Re-add Null pagination adapter for BC

This patch re-adds the `Null` pagination adapter for BC. It now emits an
`E_USER_DEPRECATED` notice on instantiation.

weierophinney committed:
Renamed Null mail transport to InMemory

`null` will be a reserved keyword in PHP 7. This patch renames the `Null` mail
transport to `InMemory`, and updates the factory to return an `InMemory`
instance whenever `null` is retrieved. A future patch will re-add the `Null`
transport with an `E_USER_DEPRECATION` notice.

weierophinney committed:
Re-add Null mail transport for BC

This patch re-adds the `Null` mail transport to keep BC. The class
extends `InMemory` and raises an `E_USER_DEPRECATED` on instantiation.

weierophinney committed:
Update paginator plugin manager to use NullFill

Added aliases for "null" and "Zend\Paginator\Adapter\Null" that point to
"nullfill", and defined that as an invokable class pointing at
`Zend\Paginator\Adapter\NullFill`.

weierophinney committed:
Updated Zend\Di test to use ToNull filter

instead of `Null` filter, to ensure tests work going forward.

Also fixed the deprecation notice in the `null` filter to point to the
appropriate replacement class.

weierophinney committed:
CS fixes as reported by php-cs-fixer

gianarb commented:
:+1: Thanks!!
